cmd/snap-bootstrap/bootstrap/tpm.go

```diff
@@ -0,0 +1,205 @@
+	"github.com/snapcore/secboot"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+const (
+	// Handles are in the block reserved for owner objects (0x01800000 - 0x01bfffff)
+	pinHandle = 0x01800000
+)
+
+type tpmSupport struct {
+	// Connection to the TPM device
+	tconn *secboot.TPMConnection
+	// Lockout authorization
+	lockoutAuth []byte
+	// List of paths to shim files
+	shimFiles []string
+	// List of paths to bootloader files
+	bootloaderFiles []string
+	// List of paths to kernel files
+	kernelFiles []string
+}
+
+func NewTPMSupport() (*tpmSupport, error) {
+	tconn, err := secboot.ConnectToDefaultTPM()
+	if err != nil {
+		return nil, fmt.Errorf("cannot connect to default TPM: %v", err)
+	}
+
+	lockoutAuth := make([]byte, 16)
+	// crypto rand is protected against short reads
+	_, err = rand.Read(lockoutAuth)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create lockout authorization: %v", err)
+	}
+
+	t := &tpmSupport{tconn: tconn, lockoutAuth: lockoutAuth}
+
+	return t, nil
+}
+
+// StoreLockoutAuth saves the lockout authorization data in a file at the
+// path specified by filename.
+func (t *tpmSupport) StoreLockoutAuth(filename string) error {
+	if err := ioutil.WriteFile(filename, t.lockoutAuth, 0600); err != nil {
+		return err
+	}
+	return nil
+}
+
+// SetShimFiles verifies and sets the list of shim binaries.
+func (t *tpmSupport) SetShimFiles(pathList ...string) error {
+	if err := ensureFilesExist(pathList); err != nil {
+		return err
+	}
+	t.shimFiles = pathList
+	return nil
+}
+
+// SetBootloaderFiles verifies and sets the list of bootloader binaries.
+func (t *tpmSupport) SetBootloaderFiles(pathList ...string) error {
+	if err := ensureFilesExist(pathList); err != nil {
+		return err
+	}
+	t.bootloaderFiles = pathList
+	return nil
+}
+
+// SetKernelFiles verifies and sets the list of kernel binaries.
+func (t *tpmSupport) SetKernelFiles(pathList ...string) error {
+	if err := ensureFilesExist(pathList); err != nil {
+		return err
+	}
+	t.kernelFiles = pathList
+	return nil
+}
+
+func ensureFilesExist(pathList []string) error {
+	for _, p := range pathList {
+		if !osutil.FileExists(p) {
+			return fmt.Errorf("file %s does not exist", p)
+		}
+	}
+	return nil
+}
+
+var (
+	provisionTPM = secboot.ProvisionTPM
+)
+
+// Provision tries to clear and provision the TPM.
+func (t *tpmSupport) Provision() error {
+	if err := provisionTPM(t.tconn, secboot.ProvisionModeFull, t.lockoutAuth); err != nil {
+		return fmt.Errorf("cannot provision TPM: %v", err)
+	}
+
+	return nil
+}
+
+// Close closes the TPM connection.
+func (t *tpmSupport) Close() error {
+	if err := t.tconn.Close(); err != nil {
+		return fmt.Errorf("cannot close TPM connection: %v", err)
+	}
+	return nil
+}
+
+var kernelCmdlines = []string{
+	// run mode
+	"console=ttyS0 console=tty1 panic=-1 systemd.gpt_auto=0 init=/sbin/init snapd_recovery_mode=run",
+	// recover mode
+	"console=ttyS0 console=tty1 panic=-1 systemd.gpt_auto=0 init=/sbin/init snapd_recovery_mode=recover",
```

Comment:
we probably don't want to add this yet

Reply:
Will keep only run mode after refactoring the cmdline setting mechanism.